### PR TITLE
Use Kokkos::abort() where necessary.

### LINKED ISSUE #16553: Use Kokkos::abort() in DEAL_II_NOT_IMPLEMENTED().
This is a necessary follow-up to https://github.com/dealii/dealii/pull/16543: As mentioned in https://github.com/dealii/dealii/pull/16543#discussion_r1468146758, we can't use `std::abort()` on devices. Instead, we must use `Kokkos::abort("text");`.

For reference, `Assert` is currently defined in the following way:
```
#  if KOKKOS_VERSION >= 30600
#    ifdef DEAL_II_HAVE_BUILTIN_EXPECT
#      define Assert(cond, exc)                                                \
        do                                                                     \
          {                                                                    \
            KOKKOS_IF_ON_HOST(({                                               \
              if (__builtin_expect(!(cond), false))                            \
                ::dealii::deal_II_exceptions::internals::issue_error_noreturn( \
                  ::dealii::deal_II_exceptions::internals::ExceptionHandling:: \
                    abort_or_throw_on_exception,                               \
                  __FILE__,                                                    \
                  __LINE__,                                                    \
                  __PRETTY_FUNCTION__,                                         \
                  #cond,                                                       \
                  #exc,                                                        \
                  exc);                                                        \
            }))                                                                \
            KOKKOS_IF_ON_DEVICE(({                                             \
              if (!(cond))                                                     \
                Kokkos::abort(#cond);                                          \
            }))                                                                \
          }                                                                    \
        while (false)
```

### PR DESCRIPTION
**Pull request by bangerth**
@masterleinad I must admit that I don't know the intricacies of what is or isn't available in Kokkos for specific versions. Does this seem right to address your comment that we can't call straight `std::abort()` in our assertion machinery?

If so, this fixes #16553.

### CODE REVIEW

**Review by masterleinad — COMMENTED** [review]
You should be able to just call `Kokkos::abort` without distinguishing between host and device execution if you were only calling a plain `std::abort` before. We only have that separation in the `Assert` macro since we want to pass through more information that is not device-compatible.

File: source/base/exceptions.cc [hunk 1]
```diff
@@ -516,7 +516,21 @@ namespace deal_II_exceptions
...
+          "Abort() was called during dealing with an assertion or exception.");
+      }))
+#else /*if KOKKOS_VERSION >= 30600*/
+#  ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
       std::abort();
+#  else
+      Kokkos::abort(
+        "Abort() was called during dealing with an assertion or exception.");
+#  endif
+#endif
```
Comment (masterleinad) [inline source/base/exceptions.cc:533]:
```suggestion
      Kokkos::abort();
```

**Comment by bangerth** [comment]
Even better, of course!

**Comment by luca-heltai** [comment]
I think you need to add a (""):

```
/__w/dealii/dealii/source/base/exceptions.cc:519:20: error: too few arguments to function ‘void Kokkos::abort(const char*)’
  519 |       Kokkos::abort();
      |       ~~~~~~~~~~~~~^~
In file included from /__w/dealii/dealii/bundled/kokkos-3.7.00/core/src/Kokkos_Core_fwd.hpp:57,
                 from /__w/dealii/dealii/bundled/kokkos-3.7.00/core/src/Kokkos_Core.hpp:55,
                 from /__w/dealii/dealii/include/deal.II/base/exceptions.h:25,
                 from /__w/dealii/dealii/source/base/exceptions.cc:15:
/__w/dealii/dealii/bundled/kokkos-3.7.00/core/src/impl/Kokkos_Error.hpp:230:56: note: declared here
  230 | KOKKOS_IMPL_ABORT_NORETURN KOKKOS_INLINE_FUNCTION void abort(
      |                                                        ^~~~~
/__w/dealii/dealii/source/base/exceptions.cc:520:5: error: ‘noreturn’ function does return [-Werror]
  520 |     }

```

**Comment by masterleinad** [comment]
> I think you need to add a (""):

Yes, you need to provide an error string.

**Comment by bangerth** [comment]
Yes, I noticed too yesterday. I didn't see it because I merged @masterleinad suggestion on the web site, rather than making the change locally and compiling. I will get to fix this later today.

**Comment by bangerth** [comment]
OK, works now!

**Review by tjhei — APPROVED** [review]

**Comment by tjhei** [comment]
I think it would make sense to leave a comment about the necessity to use the Kokkos version on device code but that it is equivalent to std::abort otherwise.

**Comment by bangerth** [comment]
I added the comment.

**Review by masterleinad — APPROVED** [review]

**Pull request merged**